libs/core/langchain_core/agents.py

```diff
@@ -189,10 +189,18 @@ def _convert_agent_observation_to_messages(
     Returns:
         AIMessage that corresponds to the original tool invocation.
     """
+
     if isinstance(agent_action, AgentActionMessageLog):
         return [_create_function_message(agent_action, observation)]
     else:
-        return [HumanMessage(content=observation)]
+        if not isinstance(observation, str):
+            try:
+                content = json.dumps(observation, ensure_ascii=False)
```

Comment:
Could you add a comment to explain why json.dumps is chosen as the default strategy vs. str? ( I'm actually not sure which one should be used -- is there a good way to think about this?)

Reply:
okay let me think about this.

Reply:
fwiw i added these gists to repro the issue

https://github.com/langchain-ai/langchain/issues/26763#issuecomment-2367315600

this issue happens with tools with dict when agent was created with create_react_agent https://gist.github.com/sharrajesh/765c0b6edfe991363675f45d467e3c93

but does not happen with other variant of agent creation create_openai_tools_agent which i cannot change without introducing risk
https://gist.github.com/sharrajesh/1080af5a95ae9d7b83a8da46597b68e1

i can give you branch a try to make sure it really works if it helps?

Reply:
Yes don't worry @sharrajesh all of these will get resolved.

Reply:
Team, when you do expect this fix will be merged?